golang/cosmos/x/swingset/keeper/extension_snapshotter.go

```diff
@@ -0,0 +1,270 @@
+// boundaries.
+func (snapshotter *SwingsetSnapshotter) InitiateSnapshot(height int64) error {
+	if !snapshotter.isConfigured() {
+		return fmt.Errorf("snapshot manager not configured")
+	}
+
+	return snapshotter.swingStoreSnapshotter.InitiateSnapshot(uint64(height), snapshotter, SwingStoreExportOptions{
+		ExportMode:        "current",
+		IncludeExportData: false,
+	})
+}
+
+// TakeSnapshot performs the app snapshot, which will call `SnapshotExtension`
+// This operation is invoked by the SwingStoreSnapshotter in a goroutine
+// started by `InitiateSnapshot`, only if no other snapshot operation is
+// already in progress.
+//
+// Implements SwingStoreSnapshotTaker
+func (snapshotter *SwingsetSnapshotter) TakeSnapshot(height uint64, retrieveSwingStoreExport func() error) error {
+	logger := snapshotter.logger.With("height", height)
+
+	// We assume SwingStoreSnapshotter correctly guarded against concurrent snapshots
+	active := activeExtensionSnapshot{
+		height:                   height,
+		logger:                   logger,
+		retrieveSwingStoreExport: retrieveSwingStoreExport,
+	}
+	snapshotter.activeExtensionSnapshot = &active
+
+	snapshotter.takeAppSnapshot(int64(height))
+
+	snapshotter.activeExtensionSnapshot = nil
+
+	// Unfortunately Cosmos `Snapshot()` does not report its errors.
+	return nil
+}
+
+// SnapshotExtension writes extension payloads into the underlying protobuf stream.
+// This operation is invoked by the snapshot manager in the goroutine started by
+// the SwingStoreSnapshotter, during the call to `TakeSnapshot`, and delegates
+// actually writing extension payloads to `ProcessExportedSnapshot`.
+// Implements ExtensionSnapshotter
+func (snapshotter *SwingsetSnapshotter) SnapshotExtension(height uint64, payloadWriter snapshots.ExtensionPayloadWriter) error {
+	logError := func(err error) error {
+		// The cosmos layers do a poor job of reporting errors, however
+		// `SwingStoreSnapshotter` arranges to report retrieve errors swallowed by
+		// `takeAppSnapshot`, so we manually report unexpected errors.
+		snapshotter.logger.Error("swingset snapshot extension failed", "err", err)
+		return err
+	}
+
+	activeExtensionSnapshot := snapshotter.activeExtensionSnapshot
+	if activeExtensionSnapshot == nil {
+		// shouldn't happen, but return an error if it does
+		return logError(errors.New("no active swingset snapshot"))
+	}
+
+	if activeExtensionSnapshot.height != height {
+		return logError(fmt.Errorf("swingset extension snapshot requested for unexpected height %d (expected %d)", height, activeExtensionSnapshot.height))
+	}
+
+	activeExtensionSnapshot.payloadWriter = payloadWriter
+
+	return activeExtensionSnapshot.retrieveSwingStoreExport()
+}
+
+// ProcessExportedSnapshot handles the snapshot retrieved by the SwingStoreSnappshotter
+// and writes it out to the SnapshotExtension's payloadWriter.
+// This operation is invoked by the SwingStoreSnapshotter nested in the goroutine
+// started by `InitiateSnapshot`.
```

Comment:
I would really appreciate an all-in-one-place explanation of how these functions relate to each other in practice.